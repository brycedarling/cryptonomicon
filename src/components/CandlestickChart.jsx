import React, { Component } from 'react';
import { connect } from 'react-redux';
import { scaleLinear, scaleTime } from 'd3-scale';
import { extent, min, max } from 'd3-array';
import Margin from './Margin';
import Axis from './Axis';
import Line from './Line';

class CandlestickChart extends Component {
  get outerWidth() {
    const margin = this.props.margin;
    return this.props.width + margin.left + margin.right;
  }

  get outerHeight() {
    const margin = this.props.margin;
    return this.props.height + margin.top + margin.bottom;
  }

  get innerWidth() {
    const margin = this.props.margin;
    return this.props.width - margin.left - margin.right;
  }

  get innerHeight() {
    const margin = this.props.margin;
    return this.props.height - margin.top - margin.bottom;
  }

  get xScale() {
    return scaleTime()
      .domain(extent(this.props.data, d => d.date))
      .range([0, this.innerWidth]);
  }

  get yScale() {
    return scaleLinear()
      .domain([
        min(this.props.data, d => d.low),
        max(this.props.data, d => d.high)
      ])
      .range([this.innerHeight, 0]);
  }

  render() {
    const margin = this.props.margin,
          outerWidth = this.outerWidth,
          outerHeight = this.outerHeight,
          innerWidth = this.innerWidth,
          innerHeight = this.innerHeight,
          xScale = this.xScale,
          yScale = this.yScale;

    return (
      <svg
        className="candlestick-chart"
        width={outerWidth}
        height={outerHeight}
      >
        <Margin
          top={margin.top}
          left={margin.left}
          bottom={margin.bottom}
          right={margin.right}
        >
          <Axis
            label={{text: 'Date', transform: `translate(${innerWidth / 2}, ${margin.bottom+5})`}}
            className="x-axis"
            orientation="bottom"
            scale={xScale}
            transform={`translate(0, ${innerHeight})`}
          />
          <Axis
            label={{text: 'Price', transform: `rotate(-90) translate(${-innerHeight / 2},${-margin.left+15})`}}
            className="y-axis"
            orientation="left"
            scale={yScale}
          />
          <Line
            x={d => xScale(d.date)}
            y={d => yScale(d.close)}
            data={this.props.data}
          />
          <Line
            x={d => xScale(d.date)}
            y={d => yScale(d.open)}
            data={this.props.data}
            stroke="green"
          />
          <Line
            x={d => xScale(d.date)}
            y={d => yScale(d.high)}
            data={this.props.data}
            stroke="red"
          />
          <Line
            x={d => xScale(d.date)}
            y={d => yScale(d.low)}
            data={this.props.data}
            stroke="yellow"
          />
        </Margin>
      </svg>
    );
  }
}

const mapStateToProps = (state) => ({
  data: state.ticks.data,
});

const mapDispatchToProps = {};

export default connect(mapStateToProps, mapDispatchToProps)(CandlestickChart);
