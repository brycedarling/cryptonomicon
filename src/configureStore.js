import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './reducers';

const configureStore = () => {
  const initialState = {};

  const enhancers = [];

  const middleware = [
    thunk,
  ];

  if (process.env.NODE_ENV === 'development') {
    const { __REDUX_DEVTOOLS_EXTENSION__ } = window;

    if (typeof devToolsExtension === 'function') {
      enhancers.push(__REDUX_DEVTOOLS_EXTENSION__());
    }
  }

  const composedEnhancers = compose(
    applyMiddleware(...middleware),
    ...enhancers,
  );

  const store = createStore(
    rootReducer,
    initialState,
    composedEnhancers,
  );

  return store;
};

export default configureStore;
