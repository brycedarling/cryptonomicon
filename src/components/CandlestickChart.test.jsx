import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import CandlestickChart from './CandlestickChart';
import configureStore from '../configureStore';

describe('components', () => {
  describe('CandlestickChart', () => {
    it('renders correctly', () => {
      const store = configureStore();

      const tree = renderer.create((
        <Provider store={store}>
          <CandlestickChart
            width={600}
            height={270}
            margin={{
              top: 30, right: 20, bottom: 30, left: 50,
            }}
          />
        </Provider>
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
