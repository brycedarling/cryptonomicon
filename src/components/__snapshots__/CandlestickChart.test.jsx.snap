// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components CandlestickChart renders correctly 1`] = `
<svg
  className="candlestick-chart"
  height={330}
  width={670}
>
  <g
    className="margin"
    transform="translate(50, 30)"
  >
    <g
      className="axis x-axis"
      transform="translate(0, 210)"
    />
    <g
      className="axis y-axis"
      transform=""
    />
    <path
      className="line"
      d={null}
      style={
        Object {
          "fill": "none",
          "shapeRendering": "crispEdges",
          "stroke": "steelblue",
          "strokeWidth": 1,
        }
      }
    />
    <path
      className="line"
      d={null}
      style={
        Object {
          "fill": "none",
          "shapeRendering": "crispEdges",
          "stroke": "green",
          "strokeWidth": 1,
        }
      }
    />
    <path
      className="line"
      d={null}
      style={
        Object {
          "fill": "none",
          "shapeRendering": "crispEdges",
          "stroke": "red",
          "strokeWidth": 1,
        }
      }
    />
    <path
      className="line"
      d={null}
      style={
        Object {
          "fill": "none",
          "shapeRendering": "crispEdges",
          "stroke": "yellow",
          "strokeWidth": 1,
        }
      }
    />
  </g>
</svg>
`;
